### PR TITLE
Aoss 1426 approve agent request

### PR DESCRIPTION
**Pull request by bravesirrobin81**

### CODE REVIEW

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala [hunk 1]
```diff
@@ -37,16 +42,25 @@ class ClientAuthorisationStatusController(override val authConnector: AuthConnec
     Ok(views.html.client.info())
   }
 
-  val showApprove = AuthorisedFor(SaClientRegime, GGConfidence) { implicit authContext => implicit request =>
-    Ok(views.html.client.approveRequest())
+  val showApprove = AuthorisedFor(SaClientRegime, GGConfidence).async { implicit authContext => implicit request =>
+    agentClientAuthorisation.requestsForAgent(null).map {
```
Comment (szmg) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala]: Will this not get all requests and pick the first one? (based on mongo index I guess, but not clientRegimeId)

File: app/uk/gov/hmrc/agentclientauthorisation/views/client/acceptRequest.scala.html [hunk 2]
```diff
@@ -1,10 +1,13 @@
-@()(implicit request: Request[_])
+@import uk.gov.hmrc.agentclientauthorisation.model.AgentClientAuthorisationRequest
+@(authRequest: Option[AgentClientAuthorisationRequest])(implicit request: Request[_])
```
Comment (szmg) [inline app/uk/gov/hmrc/agentclientauthorisation/views/client/acceptRequest.scala.html]: Why is `authRequest` optional?

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala [hunk 3]
```diff
@@ -37,16 +42,25 @@ class ClientAuthorisationStatusController(override val authConnector: AuthConnec
...
+    acceptForm match {
+      case AcceptForm(requestId, "accept") => showConfirmation(requestId, "accept")
+      case AcceptForm(_, "reject") => Future successful Ok(views.html.client.rejectRequest())
+      case _ => Future successful BadRequest
     }
+  }
+
+  private def showConfirmation(requestId: String, action: String)(implicit hc: HeaderCarrier, request: Request[_]): Future[Result] = {
+    agentClientAuthorisation.changeRequestStatus(requestId, action)
+      .flatMap(_ => agentClientAuthorisation.getRequest(requestId))
```
Comment (szmg) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala]: Should we not give back the updated instance after update instead? It seems to be a small change on the backend as the repository already supports it.

File: app/uk/gov/hmrc/agentclientauthorisation/connectors/AgentClientAuthorisationConnector.scala [hunk 4]
```diff
@@ -43,9 +41,12 @@ class AgentClientAuthorisationConnector(baseUrl: URL, http: HttpGet with HttpPos
     }
   }
 
-  // TODO
-  def requestsForClient(saUtr: SaUtr): List[AgentClientAuthorisationRequest] = {
-    List(AgentClientAuthorisationRequest(UUID.randomUUID().toString, AgentCode("AG1234"), clientRegimeId = saUtr.value, clientFullName = "Mr A Person", events = List(RequestStatusChange(Pending, new DateTime()))))
+  def changeRequestStatus(requestId: String, status: String)(implicit hc: HeaderCarrier) = {
+    http.POSTEmpty(url(s"requests/$requestId/$status"))
```
Comment (szmg) [inline app/uk/gov/hmrc/agentclientauthorisation/connectors/AgentClientAuthorisationConnector.scala]: `requestId` is coming from a user form here, shouldn't it be sanitised and/or validated on the form?

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala (hunk 1 above)
Comment (bravesirrobin81) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala]: Correct. I haven't yet considered the multiple request case.

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala (hunk 3 above)
Comment (bravesirrobin81) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala]: It'd be useful for the API to do so; but I think this should do a redirect after the post, so it'll need to re-fetch the record.

File: app/uk/gov/hmrc/agentclientauthorisation/connectors/AgentClientAuthorisationConnector.scala (hunk 4 above)
Comment (bravesirrobin81) [inline app/uk/gov/hmrc/agentclientauthorisation/connectors/AgentClientAuthorisationConnector.scala]: It should be. I think we need to consider this more generally across the entire app.

File: app/uk/gov/hmrc/agentclientauthorisation/views/client/acceptRequest.scala.html (hunk 2 above)
Comment (bravesirrobin81) [inline app/uk/gov/hmrc/agentclientauthorisation/views/client/acceptRequest.scala.html]: Can't remember my thinking here, so it's no longer optional.

File: app/uk/gov/hmrc/agentclientauthorisation/connectors/AgentClientAuthorisationConnector.scala (hunk 4 above)
Comment (frankoid) [inline app/uk/gov/hmrc/agentclientauthorisation/connectors/AgentClientAuthorisationConnector.scala]: I think that as a starter we should:
- ensure that parameters such as `requestId` are URL-path-encoded to protect against path traversal attacks (e.g. `../` in the requestId)
- in the backend service check that the current user is allowed to view the specified request and return a 403 if not

File: test/uk/gov/hmrc/agentclientauthorisation/acceptance/page/AcceptRequestPage.scala [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import uk.gov.hmrc.agentclientauthorisation.acceptance.support.AbstractWebPage
+
+class AcceptRequestPage(implicit webDriver: WebDriver) extends AbstractWebPage with ShouldMatchers {
+  override def checkIsLoadedLocator = By.className("client-authorisation-accept-page")
+
+  def agentName = find(id("agentName"))
+  def agentFriendlyName = find(id("friendlyName"))
+
+
+  def hasAgentDetails(agentName: String, agentFriendlyName: String) = {
```
Comment (szmg) [inline test/uk/gov/hmrc/agentclientauthorisation/acceptance/page/AcceptRequestPage.scala]: Could you please rename `hasSomething(...)` methods to `shouldHaveSomething(...)`? It's less ambiguous, at least for me.

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala (hunk 1 above)
Comment (szmg) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/client/ClientAuthorisationStatusController.scala]: Ah, I thought we're in the backend and it got all the requests, not only the ones that belong to the client. Sorry, you're right, multiple request case should not be considered for now.

**Pull request merged**